doc/optbot.1
.TH OPTBOT 1 LOCAL
.SH NAME
optbot - parse command line options

.SH SYNOPSIS
.B optbot [ -a
.I argument_description
.B ] --
.I given_arguments

.SH DESCRIPTION
.B optbot
allows parsing of command line options given to a shell script.  It does this
by utilizing
.BR liboptbot (1)
to handle argument descriptions given with
.B -a
options.  Options given after the
.B --
flag are treated as parameters to be parsed according to the given argument
descriptions.  If the given parameters are acceptable, assignments are
output to STDOUT as shell code for evaluation in the context in which
they will be used.  Error messages are printed to STDERR for any parsing
failures.

.IP given_arguments
These are the flags passed to the calling script.  These will usually be given
via
.B $@
\.

.SH OPTIONS

.IP "-a, --add_arg argument_description"
Add an argument descriptor.  This should be a comma seperated list of
four elements.
.IP 1.
The short option name
.IP 2.
The long option name
.IP 3.
The description of the option. Used for generating help output.
.IP 4.
"true" or "false" value,
determining whether or not this option takes values
.IP
Any of the first three options may be left blank.

.IP "-p, --prefix varaible_prefix"
The prefix that should be prepended to the variables assigned in this
script's output.  Must be shorter than 16 characters.

.IP "-h, --help"
Prints help

.IP "-d, --delimiter"
The character used to seperate arguments to --add-arg.  "," by default.

.IP "-f, --fail_var"
Prints <prefix>_success=false; on failure.  Can be used to check the success of
the script if -e is not specified.

.IP "-e, --exit_on_failure"
Causes an exit command to be printed on failure.  Should be used to cease
termination of the calling script with nonzero status.

.SH EXAMPLES

.IP
.B eval `optbot -av,verbose,"Should this script run in verbose mode?",false -- -v`
.IP
This will set $arg_verbose to true in the calling script.

.IP
.B eval `optbot -af,file,"Add an output file",true -- -f one.txt --file two.txt`
.IP
This will set $arg_file to the first given argument, one.txt in this case.
$arg_file_n will be set to the number of given arguments, 2 in this case.
$arg_file_1 and $arg_file_2 will be set to one.txt and two.txt, respectively.


src/optbot.c
#include <liboptbot.h>
#include <string.h>
#include <stdbool.h>
#include <stdlib.h>
#include <stdio.h>

struct arg_params {
  char little;
  char* big;
  char* description;
  bool takes_value;
};

/* If any function returns failure, its error message should be kept here. */
char err_msg[256];

struct arg_params* init_arg_params(void) {
  struct arg_params* params = malloc(sizeof(struct arg_params));

  params->little = '\0';
  params->big = NULL;
  params->description=NULL;
  params->takes_value = false;

  return params;
}

#define validate(condition, err_str, ...) \
  if(! (condition)) { \
    sprintf((char*) &err_msg, err_str, ##__VA_ARGS__); \
    return NULL;\
  }

struct arg_params* parse_arg_params(char* csv_string) {
  char* str;
  struct arg_params* arg = init_arg_params();

  str = strtok(csv_string, ",");
  validate(strlen(str) == 1,
    "Single character expected for little opt, \"%s\" was given.", str);
  arg->little = str[0];

  str = strtok(NULL, ",");
  validate(strlen(str) >= 2,
    "String longer than 1 char expected for big opt, \"%s\" was given.", str);
  validate(strlen(str) <= 64,
    "big opt may not be longer than 64 characters");
  arg->big = str;

  arg->description = strtok(NULL, ",");

  str = strtok(NULL, ",");
  if(strcmp("true", str) == 0) {
    arg->takes_value = true;
  } else {
    validate(strcmp("false", str) == 0, "Value expectation must be "
      "\"true\" or \"false\". \"%s\" was given.", str);
  }

  return arg;
}

void export_arg(struct cli_arg* arg, char* prefix) {
  int i;

  if(!prefix) prefix = "arg_";

  if(arg->values_length > 0) {
    printf("%s%s=%s;", prefix, arg->big, arg->values[0]);
  } else if(arg->times_set > 0){
    printf("%s%s=true;", prefix, arg->big);
  } else {
    printf("%s%s=false;", prefix, arg->big);
  }

  if(arg->values_length > 1) {
    printf("%s%s_n=%d;", prefix, arg->big, arg->values_length);
    for(i = 0; i < arg->values_length; i++) {
      printf("%s%s_%d=%s;", prefix, arg->big, i + 1, arg->values[i]);
    }
  }
}

int main(int argc, const char** argv) {
  struct cli_arg_list* arg_list = init_cli_arg_list();
  struct cli_arg_list* parse_list = init_cli_arg_list();
  struct cli_arg* arg;
  struct arg_params** argp;
  bool fail_loudly = false;
  bool fail_hard = false;
  char* prefix = NULL;
  int i;

  add_arg(arg_list, 'a', "add_arg",
    "Adds an argument to be parsed.\n\n"
    "    A comma value list should be given to each instance of this option\n"
    "    the values are as follows: little,big,description,takes_value.\n"
    "      little: The single character option for the parameter\n"
    "      big: The long option for the parameter.  this is also used as the\n"
    "           name of the environment variable that will be set if this\n"
    "           parameter is given\n"
    "      description:  The description for the argument that should show up\n"
    "                    in the help display\n\n"
    "      takes_value:  Must be one of \"true\" or \"false\".  Determines\n"
    "                    whether or not optbot will expect a value to be \n"
    "                    given for this option.  Any values given to the \n"
    "                    option will be stored in the parameter names with \n"
    "                    the big option string, with a number appended for\n"
    "                    each successive value.",
    true
  );

  add_arg(arg_list, 'p', "variable_prefix",
    "The prefix that should be appended to exported env variables\n"
    "    May not be longer than 16 characters.",
    true
  );

  add_arg(arg_list, 'h', "help", "Prints this help", false);
  add_arg(arg_list, 'd', "delimiter",
    "The character used to seperate arguments to --add-arg",
    false
  );
  add_arg(arg_list, 'f', "fail_var",
    "Prints arg_success=false; on failure.  Should be used when output\n"
    "   of this program is being eval'd.",
    false
  );

  add_arg(arg_list, 'e', "exit_on_failure",
    "Causes an exit command to be printed on failure.  Should be used when\n"
    "    output of this program is being eval'd.",
    false
  );

  /* Note that this option is provided on the parse_list */
  add_arg(parse_list, 'h', "help", "Show this help page", false);

  arg_list->devour_flag = true;

  if(! parse_command_line(arg_list, argc, argv)) {
    fprintf(stderr, "optbot failure: %s\n", arg_list->message);
    goto error;
  }

  if(little_opt_arg(arg_list, 'h')->times_set) {
    print_help(arg_list);
    goto error;
  }

  if(little_opt_arg(arg_list, 'f')->times_set)
    fail_loudly = true;

  if(little_opt_arg(arg_list, 'e')->times_set)
    fail_hard = true;

  arg = little_opt_arg(arg_list, 'p');
  if(arg->times_set > 1) {
    fprintf(stderr, "optbot failure: Only one variable prefix is allowed!\n");
    goto error;
  } else if(arg->times_set == 1) {
    prefix = strdup(arg->values[0]);
    checkmem(prefix);
  }

  arg = little_opt_arg(arg_list, 'a');
  argp = malloc(sizeof(struct arg_params*) * arg->values_length);
  checkmem(argp)
  for(i = 0; i < arg->values_length; i++) {
    argp[i] = parse_arg_params(arg->values[i]);

    if(!argp[i]) {
      fprintf(stderr, "%s\n", err_msg);
      goto error;
    }

    add_arg(parse_list, argp[i]->little, argp[i]->big,
      argp[i]->description, argp[i]->takes_value);
  }

  if(! parse_command_line(parse_list, arg_list->argc,
    (const char**)arg_list->argv))
  {
    fprintf(stderr, "%s\n", parse_list->message);
    goto error;
  }

  /* Make it possible to print the help for the script calling this by
   * passing it -h or --help */
  if(little_opt_arg(parse_list, 'h')->times_set > 0) {
    print_help(parse_list);
    goto error;
  }

  for(i = 0; i < arg->values_length; i++)
    export_arg(little_opt_arg(parse_list, argp[i]->little), prefix);

  return 0;

  error:
    if(fail_loudly) printf("arg_success=false;");
    if(fail_hard) printf("exit 1;");
    return false;
}
